photon-client/src/main/scala/com/linkedin/photon/ml/hyperparameter/ShrinkSearchRange.scala

```diff
@@ -0,0 +1,126 @@
+    val (overallPoints, overallEvals) = hyperparameterRescaled.map(x => (x._1.asDenseMatrix, DenseVector(x._2))).reduce(
+      (a, b) => (DenseMatrix.vertcat(a._1, b._1), DenseVector.vertcat(a._2, b._2))
+    )
+
+    // Fit Gaussian process regression model
+    val estimator = new GaussianProcessEstimator(kernel = new Matern52)
+    val model = estimator.fit(overallPoints, overallEvals)
+
+    // Sobol generator
+    val paramDistributions = {
+      val sobol = new SobolSequenceGenerator(numParams)
+      sobol.skipTo((seed % (Int.MaxValue.toLong + 1)).toInt)
+      sobol
+    }
+
+    // Draw candidates from a Sobol generator
+    val candidates = (1 until 1000).foldLeft(DenseMatrix(paramDistributions.nextVector)) { case (acc, _) =>
+      DenseMatrix.vertcat(acc, DenseMatrix(paramDistributions.nextVector))
+    }
+
+    // Select the best candidate
+    val predictions = model.predict(candidates)
+    val bestCandidate = selectBestCandidate(candidates, predictions._1)
+
+    // compute lower bound and upper bound
+    val upperBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => min(x + radius, 1.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+    val lowerBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => max(x - radius, 0.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+
+    discreteParams.foreach{
+      case (index, _) =>
+        upperBound(index) = min(upperBound(index) + 1, ranges(index).end)
+        lowerBound(index) = max(lowerBound(index) - 1, ranges(index).start)
+    }
+
+    (lowerBound, upperBound)
+  }
+
+  /**
+    * Selects the best candidate according to the predicted values, where "best" is defined as the largest
+    *
+    * @param candidates matrix of candidates
+    * @param predictions predicted values for each candidate
+    * @return the candidate with the best value
+    */
+  private def selectBestCandidate(candidates: DenseMatrix[Double],
+                                  predictions: DenseVector[Double]): DenseVector[Double] = {
+
+    val init = (candidates(0,::).t, predictions(0))
+
+    val (selectedCandidate, _) = (1 until candidates.rows).foldLeft(init) {
+      case ((bestCandidate, bestPrediction), i) =>
+        if (predictions(i) > bestPrediction) {
+          (candidates(i,::).t, predictions(i))
+        } else {
+          (bestCandidate, bestPrediction)
+        }
+    }
+
+    selectedCandidate
+  }
+
+  /**
+    * Discretize candidates with specified indices.
+    *
+    * @param candidate candidate with values in [0, 1]
+    * @param discreteParams Map that specifies the indices of discrete parameters and their numbers of discrete values
+    * @return candidate with the specified discrete values
+    */
+  private def discretizeCandidate(candidate: DenseVector[Double],
+                                  discreteParams: Map[Int, Int]): DenseVector[Double] = {
```

Comment:
Same as above.

Reply:
discretizeCandidate is inside RandomSearch class. To initialize an instance of the class, we need to specify the evaluation function of the type T related with different models. I think this function should be a utility function so that it works for any models universally. We are in the code migration process. These functions will be in the atlas-ml eventually so that it can be directly called.